sdk-extensions/tracing-incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/trace/ExecutorServiceSpanProcessor.java

```diff
@@ -0,0 +1,280 @@
+        spanExporter, executorService, ownsExecutorService);
+  }
+
+  ExecutorServiceSpanProcessor(
+      SpanExporter spanExporter,
+      long scheduleDelayNanos,
+      int maxQueueSize,
+      int maxExportBatchSize,
+      long exporterTimeoutNanos,
+      ScheduledExecutorService executorService,
+      boolean ownsExecutorService,
+      long workerScheduleIntervalNanos) {
+    this.worker =
+        new Worker(
+            spanExporter,
+            scheduleDelayNanos,
+            maxExportBatchSize,
+            exporterTimeoutNanos,
+            new ArrayBlockingQueue<>(maxQueueSize),
+            SPAN_PROCESSOR_TYPE_LABEL,
+            SPAN_PROCESSOR_TYPE_VALUE,
+            executorService,
+            isShutdown,
+            workerScheduleIntervalNanos);
+    this.ownsExecutorService = ownsExecutorService;
+    this.executorService = executorService;
+    executorService.schedule(worker, workerScheduleIntervalNanos, TimeUnit.NANOSECONDS);
+  }
+
+  @Override
+  public void onStart(Context parentContext, ReadWriteSpan span) {}
+
+  @Override
+  public boolean isStartRequired() {
+    return false;
+  }
+
+  @Override
+  public void onEnd(ReadableSpan span) {
+    if (!span.getSpanContext().isSampled()) {
+      return;
+    }
+    worker.addSpan(span);
+  }
+
+  @Override
+  public boolean isEndRequired() {
+    return true;
+  }
+
+  @Override
+  public CompletableResultCode shutdown() {
+    if (isShutdown.getAndSet(true)) {
+      return CompletableResultCode.ofSuccess();
+    }
+    CompletableResultCode result = worker.shutdown();
+    // do the cleanup after worker finishes flush
+    result.whenComplete(
+        () -> {
+          if (ownsExecutorService) {
+            executorService.shutdown();
+          }
+        });
+
+    return result;
+  }
+
+  @Override
+  public CompletableResultCode forceFlush() {
+    return worker.forceFlush();
+  }
+
+  // Visible for testing
+  List<SpanData> getBatch() {
+    return new ArrayList<>(worker.batch);
+  }
+
+  private static class Worker implements Runnable {
+
+    private final AtomicLong nextExportTime = new AtomicLong();
+    private final ArrayBlockingQueue<SpanData> batch;
+    private final AtomicBoolean isShutdown;
+    private final long workerScheduleIntervalNanos;
+    private final WorkerExporter workerExporter;
+    private final BoundLongCounter droppedSpans;
+    private final AtomicReference<CompletableResultCode> flushRequested = new AtomicReference<>();
+    private final long scheduleDelayNanos;
+    private final BlockingQueue<ReadableSpan> queue;
+    private final int maxExportBatchSize;
+    private final SpanExporter spanExporter;
+    private final ScheduledExecutorService executorService;
+
+    private Worker(
+        SpanExporter spanExporter,
+        long scheduleDelayNanos,
+        int maxExportBatchSize,
+        long exporterTimeoutNanos,
+        BlockingQueue<ReadableSpan> queue,
+        String spanProcessorTypeLabel,
```

Comment:
Is there a good reason to have these part of the constructor? Aren't they always the same? I'd rather just have them be constants, probably in the enclosing class.